src/app/data-model-provider/MetadataTypes.h

```diff
@@ -20,16 +20,46 @@
 #include <optional>
 
 #include <access/Privilege.h>
+#include <app-common/zap-generated/cluster-objects.h>
 #include <app/ConcreteAttributePath.h>
 #include <app/ConcreteClusterPath.h>
 #include <app/ConcreteCommandPath.h>
+#include <app/data-model/List.h>
 #include <lib/core/DataModelTypes.h>
 #include <lib/support/BitFlags.h>
+#include <lib/support/BitMask.h>
 
 namespace chip {
 namespace app {
 namespace DataModel {
 
+enum class EndpointCompositionPattern : uint8_t
+{
+    kTreePattern       = 0x1,
+    kFullFamilyPattern = 0x2,
+};
+
+struct EndpointInfo
+{
+    // kInvalidEndpointId if there is no explicit parent endpoint (which means the parent is endpoint 0,
+    // for endpoints other than endpoint 0).
+    EndpointId parentId;
+    EndpointCompositionPattern compositionPattern;
+
+    explicit EndpointInfo(EndpointId parent) : parentId(parent), compositionPattern(EndpointCompositionPattern::kFullFamilyPattern)
+    {}
+    explicit EndpointInfo(EndpointId parent, EndpointCompositionPattern pattern) : parentId(parent), compositionPattern(pattern) {}
```

Comment:
since this is no default conversion (2 arguments) this probably does not need to be explicit.